### ISSUE TITLE
Cannot get MathJax-node to work in Windows

### ISSUE THREAD

**Issue by sammarshallou** [mathjax/MathJax-node#70]
I successfully got the npm install to work in Windows 7 (which was a nightmare and a half already), but the MathJax-node binaries don't work for me. Using mingw bash (also tested with windows command prompt just in case that helped, but it has the same result).

I run:

$ bin/tex2svg 'x_2'

The result is nothing - i.e. it pauses for about 1-2 seconds then exits and shows the $ prompt again, with no output. 

Originally I was using node.js v0.10.something. I've since updated to the latest v0.12.0 (and did a rebuild and update) but with same result.

I tried adding console.log calls in various places in the script (bin/tex2svg). As a result I can confirm that it runs the script from start to finish, but does not run the callback that is supposed to happen when typesetting finishes.

I also have a Linux VM where I also installed MathJax-node the same way and this does work fine, returning SVG code.

Basically wondering if this is just my setup, or if it's known to not work in Windows for some reason. Our production servers are Linux but for development it is sometimes convenient if we can run stuff on Windows also.

**Comment by pkra** [comment]
Thanks for the report. I don't think we've had anyone test MathJax-node on Windows.

Given that you've gone through it, do you have any advice on pitfalls during installation?

**Comment by sammarshallou** [comment]
Thanks! OK, I was wondering whether anybody else had tried it. Seems a bit strange that there's no error message (that I can see). It's entirely possible that this is just something wrong with my system somehow but maybe somebody else will comment if it's a shared issue.

The installation difficulty I had isn't mathjax-node's problem - it's really just because mathjax uses the node.js jsdom module, specifically I think it's 'contextify' which is a dependency of that which causes the difficulty because it is a native code thing that is compiled using the 'gyp' module.

Basically in order to get the 'npm install' command to work in my Windows 7 (64-bit) I had to to:

0. Install node.js (well, actually I had this already)
1. Install Python 2.7 (apparently later versions won't work)
2. Install MS Visual Studio 2012 Express (that's the name of the free version; ironically, this install took approximately 5 hours).
3. Set the environment variable GYP_MSVS_VERSION to 2012

After that, npm install worked. The hard part was finding the environment variable because it's not really documented that I saw, I eventually got it from somebody else's slightly different problem in a google search - the rest is covered in https://github.com/TooTallNate/node-gyp

**Comment by pkra** [comment]
Thanks for the additional information. I got mathjax-node running (though I didn't need step 4) and could reproduce this -- but ran out of time. Will try to find some more time later this week.

**Comment by dpvc** [comment]
I was trying out jsdom 3.1.1, and it failed in the same way you describe, and tracked it down to the handling of the MathJax URL, which is a `file://` URL.  It looks like the `url.resolve()` method in node doesn't handle some URL's correctly, and Windows URL's with disks appear to be one of them.  Try editing `lib/mj-single.js` and find the line that sets `MathJaxPath`.  Change the `"file://"` to just `"file:"` (no slashes) and see if that helps.  If not, add

    console.error(MathJaxPath);

right after that line and see what the path actually is and let us know.  I think it is probably that jsdom isn't able to locate MathJax.js properly.

**Comment by sammarshallou** [comment]
Thanks for looking into this! That didn't solve it but you're right about this being an error. After your change the path was:

file:c:\users\sm449\workspace\MathJax-Node\node_modules\MathJax\unpacked\MathJax.js

Which still doesn't work. After various experimentation I found a change that does work (after the original code):

    if (/^file:\/\/[A-Za-z]:\\/.test(MathJaxPath)) {
        MathJaxPath = MathJaxPath.replace(/\\/g, '/');
        MathJaxPath = MathJaxPath.replace(/^file:\/\/(.):\//, 'file:///$1:/');
    }

This makes it use the URL that Firefox accepts, which is file:///c:/whatever/etc

Not sure if this is the best way to solve the problem (I'm not at all familiar with node.js) but if you want to use that change, it's this commit:

https://github.com/sammarshallou/MathJax-node/commit/30cc46ae9156468a968c50ddc399992444c24ebe

(Feel free to do it differently in your own commit obviously. :) Or if you want me to submit it using the pull request thingy, I can.)

**Comment by dpvc** [comment]
Thanks for checking that out and finding a solution.  Can you do one more thing for me?  Can you run

    node -e 'console.log(require("jsdom").jsdom().URL)'

on the command line and tell me what the output is?

**Comment by sammarshallou** [comment]
Sure. It is:

    file:///c:/users/sm449/workspace/MathJax-Node/[eval]

**Comment by dpvc** [comment]
OK, thanks.  Now how about

    node -e 'console.log(require("url").resolve("file:///c:/blank.html","file:c:/users/sm449/test.html"))'

Hate to keep making you do this, but I haven't set up MathJax-node on Windows yet myself (though I will probably have to).

**Comment by sammarshallou** [comment]
I'm happy to do this (so long as you don't mind waiting if I'm not at my desk or don't notice it for an hour or something). :) Actually pretty glad to get it working in Windows at all now as that's going to make my life easier, so thanks.

Result to that one is a bit weird:
    file:///c:/c:/users/sm449/test.html

**Comment by dpvc** [comment]
If you are willing to keep trying, here is another couple:

    node -e 'console.log(require("url").resolve("about:blank","file:c:\\users\\sm449\\test.html"))'
    node -e 'console.log(require("url").resolve("about:blank","file://c:\\users\\sm449\\test.html"))'
    node -e 'console.log(require("url").resolve("about:blank","file:///c:\\users\\sm449\\test.html"))'
    node -e 'console.log(require("url").resolve("file:///blank.html","file:c:\\users\\sm449\\test.html"))'
    node -e 'console.log(require("url").resolve("file:///blank.html","file://c:\\users\\sm449\\test.html"))'
    node -e 'console.log(require("url").resolve("file:///blank.html","file:///c:\\users\\sm449\\test.html"))'

**Comment by dpvc** [comment]
PS, I'm going to be away from my computer for the rest of the day starting in about an hour, so we may not get this resolved today.

I'm glad you were able to get it to work for you on Windows, and that you are able to work more efficiently there.

**Comment by sammarshallou** [comment]
I'm not in any hurry. Here's the results from the 6 lines above (in order, I copy/pasted it all):

    $ node -e 'console.log(require("url").resolve("about:blank","file:c:\\users\\sm449\\test.html"))'
    file://c:/users/sm449/test.html
    
    $ node -e 'console.log(require("url").resolve("about:blank","file://c:\\users\\sm449\\test.html"))'
    file://c/users/sm449/test.html
    
    $ node -e 'console.log(require("url").resolve("about:blank","file:///c:\\users\\sm449\\test.html"))
    '
    file://c:/users/sm449/test.html
    
    $ node -e 'console.log(require("url").resolve("file:///blank.html","file:c:\\users\\sm449\\test.htm
    l"))'
    file:///c:/users/sm449/test.html
    
    $ node -e 'console.log(require("url").resolve("file:///blank.html","file://c:\\users\\sm449\\test.h
    tml"))'
    file://c/users/sm449/test.html
    
    $ node -e 'console.log(require("url").resolve("file:///blank.html","file:///c:\\users\\sm449\\test.
    html"))'
    file:///c:/users/sm449/test.html

**Comment by dpvc** [comment]
OK, great.  That gives two versions that work properly.

One final thing to try out.  In `lib/mj-single.js`, change the line that sets `MathJaxPath` to be

    var MathJaxPath = "file:"+require.resolve('MathJax/unpacked/MathJax');

and in the function `GetWindow` (line 111), add

    document._URL = "file:///blank.html";  // hack to get url.resolve to work with MathJaxPath

ad the bottom of the function.  Of course, you also need to remove your work-around as well, so we can see if this fixes the problem.  I'm hoping this fixes the issue, and it should work for all platforms.

**Comment by sammarshallou** [comment]
Unfortunately this does not work. I presume because require.resolve does not use the document URL. I made it display the MathJaxPath again after this change, and it's:

    file:c:\users\sm449\workspace\MathJax-Node\node_modules\MathJax\unpacked\MathJax.js

This result is identical with or without the line in GetWindow.

Based on your approach but simply creating a new URL to resolve from, I've done this:

https://github.com/sammarshallou/MathJax-node/commit/e1495ed19ee8a7b6cf4e5a9ee5c055fc75ecf423

This change actually works (on Windows). Incidentally, when I added logging to see what it was doing, I noticed that it results in the MathJaxPath being a 'path' type (just c:\whatever) instead of a 'URL' type. 

By the way, I've discovered some other potential problems when running in Windows, I'll probably submit separate pull requests for these.

**Comment by dpvc** [comment]
Thanks for checking further, but there is a slight misunderstanding.  It is not `MathJaxPath` that is the end result that needs to be checked.  The problem comes in the internals of jsdom so that even a correct value of `MathJaxPath` will be processed incorrectly.

You say that `require.resolve()` doesn't use the document URL, which is true, but that is not the call that is causing the problems.  It is a call to `url.resolve()`, a very different routine, later on as part go `node_modules/jsdom/lib/jsdom/utils.js` where the document's base URL is used as the first parameter and the `MathJaxPath` as the second that is the problem.

I'm trying to get a method that works for all platforms and am also trying to move up to jsdom 3.1.1 rather than jsdom 1.0.2 that we currently use.  The reason I had you check the `url.resolve()` calls with `about:blank` is that current versions of jsdom use that as the base URL, while 1.0.2 uses a `file://` URL (it's the one I asked you to print out near the change in labels above).

While your change does allow `MathJaxPath` to be set correctly up front, that only works in 1.0.2 because the base URL is a `file://` URL so that when `url.reslove()` is used with it later, you still get the correct URL.  But if it is used with `about:blank`, even the correct `MathJaxPath` produces the wrong result, as shown in the test with six node commands above.  None of the `about:blank` values resolves correctly, regardless of the path sent to it.  So even if `MathJaxPath` is set up correctly initially, it will not work in jsdom 3.1.1.  That is why I was trying the change in the document base URL.  That works for me in Mac OS, and I was hoping that it would work for you as well.

Just to be sure, you did include the underscore in `document._URL`, right?  It won't work without it.

**Comment by dpvc** [comment]
Just to be excruciatingly clear, 

> I made it display the MathJaxPath again after this change, and it's:
> 
>         file:c:\users\sm449\workspace\MathJax-Node\node_modules\MathJax\unpacked\MathJax.js
> 
> This result is identical with or without the line in `GetWindow`.

this is the expected behavior.  You won't be able to see the correct URL unless you were to put a `console.log()` call into the `node_modules/jsdom/lib/jsdom/utils.js` file. With the change to `GetWindow()`, when `url.resolve()` is called with the base URL of `file:///blank.html" and the `MathJaxPath` you list above, the result should be the correct URL internally.  You should be able to tell by the fact that MathJax.js will load and the `bin/tex2svg` command will produce output.

**Comment by sammarshallou** [comment]
OK, sorry I didn't understand! I did try it before as well as just displaying it, and it still doesn't work with that change, but actually (I had kind of noticed this but not realised it was important) the error is different. Rather than displaying nothing, you do get an error message. So maybe it's on the way to working?

    $ bin/tex2svg 'a^0'
    c:\users\sm449\workspace\MathJax-Node\lib\mj-single.js:311
            MathJax.OutputJax.SVG.resetGlyphs(true);
                                 ^
    TypeError: Cannot read property 'resetGlyphs' of undefined
        at c:\users\sm449\workspace\MathJax-Node\lib\mj-single.js:311:30
        at Timer.listOnTimeout (timers.js:110:15)

Here's the change I made so you can check I got it right:

https://github.com/sammarshallou/MathJax-node/commit/18c0820d38ac7adfe374a950093693f1f8201631

(note: I have a bunch of meetings next, so won't be able to get back to you immediately. Btw re the other things, happy to sign whatever needed...)

**Comment by dpvc** [comment]
Thanks for the update.  This does mean that `MathJax.js` was loaded, so that does solve the issue with the `url.resolve()` problem.  Now what I think is happening is that MathJax isn't finding its root URL correctly due to the backslashes not being slashes.  I think that gives me the information I need to resolve the problem more globally.

Have fun at the meetings!

**Comment by dpvc** [comment]
Well, I did finally install MathJax-node on Windows, and you are right, Visual Studio Express was certainly slow to install and run.  But once there, it did do the job.

I was right about the cause of the last error messages that you give above, and have taken a version of your `url.resolve()` solution as the resolution for that problem.  So with that I think version 0.3.0 is ready to go.  See pull request #71.

**Comment by pkra** [comment]
I'm closing this. The remaining issue (png generation) should be part of #59
